### ISSUE TITLE
How to detect if window is active

### ISSUE THREAD

**Issue by miguelpduarte** [itchyny/lightline.vim#380]
First and foremost, great plugin! Love the spirit of it and it's pretty awesome!

Following some of the suggestions in #22, I configured some custom colors and made the filename bold.
However, I would like to make it so that the text is not bold when the window is inactive, but could not find how to detect this in `vim` directly. I also looked through the issues here and could not find anything regarding this as well.

Is there any easy way to detect if the current window is active? (Similar to using the `mode()` function in the above issue to use the correct colors based on the current mode)

Thanks!

**Comment by itchyny** [comment]
I think it is impossible to detect this. Instead, you can register two functions for both active and inactive windows while the config looks a little messy.
```vim
let g:lightline = {
      \ 'component_function': {
      \   'something_active': 'LightlineSomethingActive',
      \   'something_inactive': 'LightlineSomethingInactive',
      \ },

function! LightlineSomethingActive()
  return LightlineSomething(1)
endfunction
function! LightlineSomethingInactive()
  return LightlineSomething(0)
endfunction
function! LightlineSomething(active)
  return a:active ? ... : ...
endfunction

**Comment by miguelpduarte** [comment]
That's actually a pretty good idea! Thanks!

I'll probably be able to use `component` and simply not include the `%##` highlight group, reusing the same functions too. Will try that tonight and report with results 😄

**Comment by miguelpduarte** [comment]
It worked! I managed to do it with the following code, leaving it here if other people have the same wish I do (irrelevant parts not show here) - show the relative path with the filename in bold but only when the window is focused + a colored modified symbol (much like the python version of `powerline`):

```vim
let g:lightline = {
      \ 'colorscheme': 'default',
      \ 'active': {
      \   'left': [['mode', 'paste'], ['gitbranch', 'readonly', 'filename+modified']],
      \   'right': [['lineinfo'], ['percent'], ['fileformat', 'fileencoding', 'filetype']],
      \ },
      \ 'inactive': {
      \   'left': [['filename+modified_inactive']],
      \   'right': [['lineinfo'], ['percent']],
      \ },
      \ 'tabline': {
      \   ...
      \ },
      \ 'tab': {
      \  ...
      \ },
      \ 'component': {
      \   'filename+modified': '%{LightlineFilepath()}%#BoldFileName#%{LightlineFilename()}%#ModifiedColor#%{LightlineModified()}',
      \   'filename+modified_inactive': '%{LightlineFilepath()}%{LightlineFilename()}%{LightlineModified()}',
      \ },
      \ 'component_function': {
      \   ...
      \ },
      \ 'tab_component': {
      \   ...
      \ },
      \ }

...

function! LightlineFilepath()
    let relativeparent = expand('%:h')
    if relativeparent !=# '.'
	return relativeparent . '/'
    else
	return ''
    endif
endfunction

function! LightlineFilename()
    " Adjusting based on mode - kinda hardcoded, but do not yet know much
    " vimscript so this works and I'll just roll with it I guess
    "
    " Using '{g:lightline.colorscheme}' makes this based on current theme and not always powerline
    " 4 length array: [guiforeground, guibackground, ctermfg, ctermbg]
    if mode() ==# 'i'
	let colors = g:lightline#colorscheme#{g:lightline.colorscheme}#palette.insert.left[1]
    else
	let colors = g:lightline#colorscheme#{g:lightline.colorscheme}#palette.normal.left[1]
    endif

    exe printf('hi BoldFileName ctermfg=%d ctermbg=%d guifg=%s guibg=%s term=bold cterm=bold',
              \ colors[2], colors[3], colors[0], colors[1])
    let filename = expand('%:t') !=# '' ? expand('%:t') : '[No Name]'
    return filename
endfunction

function! LightlineModified()
    " As per https://github.com/itchyny/lightline.vim/issues/22 (with
    " adaptations)
    
    " Getting current colors for respective left component based on mode and
    " colorscheme
    if mode()[0] ==# 'i'
	let left_colors = g:lightline#colorscheme#{g:lightline.colorscheme}#palette.insert.left[1]
    else
	let left_colors = g:lightline#colorscheme#{g:lightline.colorscheme}#palette.normal.left[1]
    endif
    
    " Getting correct background colors for term and gui based on the current
    " colors
    let gui_bgcolor = left_colors[1]
    let term_bgcolor = left_colors[3]

    " Configuring the modified color
    let gui_modifiedcolor = '#ffaf00'
    let term_modifiedcolor = 214

    exe printf('hi ModifiedColor ctermfg=%d ctermbg=%d guifg=%s guibg=%s term=bold cterm=bold',
        \ term_modifiedcolor, term_bgcolor, gui_modifiedcolor, gui_bgcolor)

    let modified = &modified ? ' +' : &modifiable ? '' : ' -'
    return modified
endfunction
```

However, I found 2 problems:

- I could not get `component_visible_condition` to work with `winwidth(0)` (the condition I inserted there was `'winwidth(0) > 90'` to hide encoding and file format when the window became too small. I thought that the problem was that the components were not being defined by me in `component` (despite being defined by lightline) - however even after copying the definition of the components from `autoload/lightline.vim` this did not work. Is this a known bug, intended, or am I doing something wrong?
- I could not manage to override tab components using `tab_component`, as the string does not seem to be expanded. I currently have `'filename': '%{LightlineTabFilepath()}%#BoldTabFileName#%{LightlineTabFilename()}'` with the functions declared below, but the string appears as is, without any substitution (i.e the literal value of this string). Am I doing something wrong? :sweat_smile:

**Comment by itchyny** [comment]
1. You cannot use `component_visible_condition` to hide the components, they are hint for controlling the subseparators. This might be confusing but this is an expected behavior so please refer to the document. In order to hide the encoding and file format, please use `component_function` (or `%{winwidth(0) < 90 ? "" : &fenc!=#""?&fenc:&enc}` as a `component`).
2. You can use `tab_component_function` instead of `tab_components`. The behavior of `tab_component` looks weird but that's how it works and cannot be changed.

**Comment by miguelpduarte** [comment]
1. I found the solution using `component_function` in the readme, and so was what I used, but was experimenting and found `component_visible_condition` and thus wanted to try it. Thanks for the feedback though, it makes more sense now.

2. And would that work with highlight groups as well? I would like to use one here, but from what I've seen, such is not possible when using component functions.

I guess I'll try it anyway :sweat_smile: 

Thanks once again! :slightly_smiling_face:
